### PR TITLE
Premint: first minter rewards

### PR DESCRIPTION
**Pull request by oveddan**

### CODE REVIEW

**Comment by oveddan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #161** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/161" target="_blank">Graphite</a> 
    * **PR #162** <a href="https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/162" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ourzora/zora-1155-contracts/162?utm_source=stack-comment).

File: src/nft/ZoraCreator1155Impl.sol [hunk 1]
```diff
@@ -444,15 +445,38 @@ contract ZoraCreator1155Impl is
...
+
         // Get value sent and handle mint rewards
-        uint256 ethValueSent = _handleRewardsAndGetValueSent(msg.value, quantity, getCreatorRewardRecipient(), createReferrals[tokenId], mintReferral);
+        uint256 ethValueSent = _handleRewardsAndGetValueSent(
+            msg.value,
+            quantity,
+            getCreatorRewardRecipient(),
+            createReferrals[tokenId],
+            mintReferral,
+            firstMinters[tokenId]
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:457]: seems like we are double looking up `firstMinters[tokenId]` here and in `_handleFirstMinter` - can this be reduced to one storage lookup?

**Review by oveddan — COMMENTED** [review]
can you please add test that shows: first minter gets rewards on multiple mints from different minters

**Comment by kulkarohan** [comment]
> can you please add test that shows: first minter gets rewards on multiple mints from different minters

https://github.com/ourzora/zora-1155-contracts/pull/162/files#diff-4351e9f1c8712bdd84b0b795fa32bc3f4b01e309cf2b3aeb21d789bef6279cd2R969-R999

File: src/nft/ZoraCreator1155Impl.sol (hunk 1 above)
Comment (kulkarohan) [reply src/nft/ZoraCreator1155Impl.sol]: https://github.com/ourzora/zora-1155-contracts/pull/162/files#diff-2a72e8508d3c8b8b46fde1dda4e0f8720176a5839bfb9c1490f38173325d2687R469-R483

File: src/nft/ZoraCreator1155Impl.sol [hunk 2]
```diff
@@ -444,15 +445,38 @@ contract ZoraCreator1155Impl is
...
 
         // Execute commands returned from minter
         _executeCommands(minter.requestMint(msg.sender, tokenId, quantity, ethValueSent, minterArguments).commands, ethValueSent, tokenId);
 
         emit Purchased(msg.sender, address(minter), tokenId, quantity, msg.value);
     }
 
+    function _handleFirstMinter(uint256 tokenId, bytes calldata data) internal {
+        if (firstMinters[tokenId] == address(0)) {
+            address rewardRecipient = abi.decode(data, (address));
```
[the remark this answers is not in the document]
Comment (kulkarohan) [reply src/nft/ZoraCreator1155Impl.sol:474]: https://github.com/ourzora/zora-1155-contracts/pull/162/files#diff-4351e9f1c8712bdd84b0b795fa32bc3f4b01e309cf2b3aeb21d789bef6279cd2R1001-R1028

File: test/minters/merkle/ZoraCreatorMerkleMinterStrategy.t.sol [hunk 3]
```diff
@@ -109,17 +114,20 @@ contract ZoraCreatorMerkleMinterStrategyTest is Test {
...
-        vm.deal(mintTo, 20 ether);
+        uint256 totalSale = 10 ether;
+        uint256 totalReward = target.computeTotalReward(10);
+        uint256 totalValue = totalSale + totalReward;
+
+        vm.deal(mintTo, totalValue);
 
         uint256 maxQuantity = 10;
-        uint256 pricePerToken = 1000000000000000000;
+        uint256 pricePerToken = 1 ether;
```
Comment (oveddan) [inline test/minters/merkle/ZoraCreatorMerkleMinterStrategy.t.sol:124]: nice change here to these units

File: test/nft/ZoraCreator1155.t.sol [hunk 4]
```diff
@@ -963,6 +966,66 @@ contract ZoraCreator1155Test is Test {
         target.mintWithRewards(fixedPriceMinter, tokenId, quantity, abi.encode(recipient), address(0));
     }
 
+    function test_FirstMinterRewardReceivedOnConsecutiveMints(uint256 quantity) public {
```
Comment (oveddan) [inline test/nft/ZoraCreator1155.t.sol:969]: awesome test!

**Review by oveddan — COMMENTED** [review]
great improvements and nice new test coverage

**Review by kulkarohan — APPROVED** [review]

File: src/nft/ZoraCreator1155Impl.sol [hunk 5]
```diff
@@ -444,15 +446,48 @@ contract ZoraCreator1155Impl is
...
+        );
 
         // Execute commands returned from minter
         _executeCommands(minter.requestMint(msg.sender, tokenId, quantity, ethValueSent, minterArguments).commands, ethValueSent, tokenId);
 
         emit Purchased(msg.sender, address(minter), tokenId, quantity, msg.value);
     }
 
+    /// @dev Get and/or set the first minter a token
+    function _handleFirstMinter(uint256 tokenId, bytes calldata data) internal returns (address) {
```
Comment (oveddan) [inline src/nft/ZoraCreator1155Impl.sol:469]: very minor (and optional improvement) - for readability, having `returns (address firstMinter)` would be clearer.

**Comment by oveddan** [comment]
Looks like test coverage went down to 69.5% - maybe some tests were deleted?

**Comment by oveddan** [comment]
todo:
* [ ] fix test coverage
* [ ] fix fork tests

**Pull request merged**